### ISSUE TITLE
Type narrowing not applied to a generator

### ISSUE THREAD

**Issue by spagh-eddie** [python/mypy#13181]
**Bug Report**

An argument typed as `None | str` gets correctly narrowed to `str` after an `if argument:`, but the narrowing is not applied inside the `if` block. 

**To Reproduce**

```python
words = ["this", "that", "other"]


def fizz(mandatory_letters: None | str):
    filters = []
    reveal_type(mandatory_letters)
    if mandatory_letters:
        reveal_type(mandatory_letters)
        filters.append(lambda word: all(c in word for c in mandatory_letters))
    return filter(lambda word: all(f(word) for f in filters), words)
```

**Expected Behavior**

I expected no error. 
I expected the type narrowing to apply to the `if` block and its generator. 

**Actual Behavior**

```console
% mypy file.py
file.py:6: note: Revealed type is "Union[None, builtins.str]"
file.py:8: note: Revealed type is "builtins.str"
file.py:9: error: Item "None" of "Optional[str]" has no attribute "__iter__" (not iterable)
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 
```console
% mypy --version
mypy 0.961 (compiled: yes)
```
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10.1
- Operating system and version: macOS Monterey 12.0.1 (21A559)

**Comment by JelleZijlstra** [comment]
This is an instance of #2608.
